File: 189.rotate-array.js
/*
 * @lc app=leetcode id=189 lang=javascript
 *
 * [189] Rotate Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function (nums, k) {
  const reverse = (i, j) => {
    while (i < j) {
      const tmp = nums[i];
      nums[i++] = nums[j];
      nums[j--] = tmp;
    }
  };
  const len = nums.length;
  if (!len || k % len === 0) {
    return;
  }
  k %= len;
  reverse(0, len - 1);
  reverse(0, k - 1);
  reverse(k, len - 1);
};
// @lc code=end

File: 189.rotate-array.py
#
# @lc app=leetcode id=189 lang=python3
#
# [189] Rotate Array
#

# Time:  O(n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        if not nums or k % len(nums) == 0:
            return

        def reverse(begin, end):
            while begin < end:
                nums[begin], nums[end - 1] = nums[end - 1], nums[begin]
                begin += 1
                end -= 1

        k %= len(nums)
        reverse(0, len(nums))
        reverse(0, k)
        reverse(k, len(nums))


# @lc code=end
